Tweak the introduction to akka persistence (#25916)
akka/akka: akka-docs/src/main/paradox/persistence.md

## Introduction

Akka persistence enables stateful actors to persist their state so that it can be recovered when an actor

> is either restarted after a JVM crash or by a supervisor, or migrated within a cluster. The key concept behind Akka

Reviewer: You could also have an app that explicitly stops and starts persistent actors when they are needed. I think `started` should stay.

Author: I see.  I'll add it back.

Author: I didn't.  But I included your stop-start scenario, and loosened the wording to indicate these are just examples of things that can lead to a restart and recovery.

> persistence is that only the _messages_ received by the actor are persisted, not the actual state of the actor

Reviewer: It should not be the messages it receives (that'd be command sourcing) it is events representing state changes.

Author: Oh right. lol

> persistence is that only the _events_ received by the actor are persisted, not the actual state of the actor

Reviewer: The actor receives commands (`ChangeTheName(newName)`) which are then validated and events (`NameChanged(newName)`) are created and those are persisted, so maybe "_events_ representing changes that happened to the actor state" or something like that?

Author: Do all events represent changes in state?

What is it you're trying to achieve with your wording?

Reviewer 2: @dwijnand I think the wording is a bit dense. How about:

```
Akka persistence enables stateful actors to persist their state so that it can be recovered when an actor is either restarted, such as after a JVM crash, by a supervisor or a manual stop-start, or migrated within a cluster. The key concept behind Akka persistence is that it's not the actor's state that is persisted but each change on the state. These state changes are called events. The events are persisted by appending to storage (no data in the storage is ever mutated) which allows for very high transaction rates and efficient replication. A stateful actor is recovered by replaying the stored events to the actor, allowing it to rebuild its state. This reconstruction operation is called rehydration. The rehydration can be either the full replay of events or starting from a snapshot (when available) dramatically reducing recovery times. 
```
I tried to focus on the concepts (e.g. removed the use of `receive`) and introduced some common language gradually (events, rehydration,...).

Finally, I don't know what the last sentence means:
```
Akka persistence also provides point-to-point communication with at-least-once message delivery semantics.
```

Author: Now you got me thinking it shouldn't say "enables stateful actors to persist their state" either...

Reviewer 2: >  "enables stateful actors to persist their state" 

Hmmm, good point. I guess we could say:

```
Akka persistence adds durability to stateful actors so that the state can 
be recovered when an actor is either restarted, such as after a JVM crash, by a supervisor or a manual stop-start, or migrated within a cluster.
```
WDYT?

Author: I think it's better.  But perhaps it's worth add a short definition of "durability".

Also, I've been thinking that it might be better (in your previous update) to lead with technical terms first and backfill the meaning.  So instead of

> A stateful actor is recovered by replaying the stored events to the actor, allowing it to rebuild its state. This reconstruction operation is called rehydration.

something like

> A stateful actor is recovered (by/with/via) rehydration, (that is?) by replaying the stored events to the actor allowing it to rebuild its state.

I'm not sure if it's a good idea.

Reviewer 2: SGTM

Reviewer: > What is it you're trying to achieve with your wording?

Get the introduction to use event sourcing terminology but still explain in a clear way what it does (and not something it doesn't do).

I think I like @ignasi35 suggestion best so far.

> Finally, I don't know what the last sentence means:
>> Akka persistence also provides point-to-point communication 
>> with at-least-once message delivery semantics.

Akka Persistence contains the `AtLeastOnceDelivery` abstraction that does what it says on top of persistence.

Reviewer: One thought though is that even if I have seen the term `rehydrating` in event sourcing contexts, the persistence APIs and existing docs use the terms `recover` the state by `replay` of the events, so maybe we should stick with those.

Author: I didn't specify which Monday.  ⌛️

> (though actor state snapshot support is also available). The messages are persisted by appending to storage (nothing is ever mutated) which

Reviewer 3: ```suggestion
(though actor state snapshot support is also available). Events are persisted to storage by an append-only log which
```

allows for very high transaction rates and efficient replication. A stateful actor is recovered by replaying the stored

> messages to the actor, allowing it to rebuild its state. This can be either the full history of changes

Reviewer 3: ```suggestion
events and allowing the actor to rebuild its state. This can be either the full history of changes
```

> or starting from a snapshot which can dramatically reduce recovery times. Akka persistence also provides point-to-point

Reviewer 3: ?
```suggestion
or starting from a checkpoint in a snapshot which can dramatically reduce recovery times. Akka persistence also provides point-to-point
```

 * @scala[`PersistentActor`]@java[`AbstractPersistentActor`]: Is a persistent, stateful actor. It is able to persist events to a journal and can react to
them in a thread-safe manner. It can be used to implement both *command* as well as *event sourced* actors.
When a persistent actor is started or restarted, journaled messages are replayed to that actor so that it can
recover its state from these messages.
 * @scala[`AtLeastOnceDelivery`]@java[`AbstractPersistentActorAtLeastOnceDelivery`]: To send messages with at-least-once delivery semantics to destinations, also in
case of sender and receiver JVM crashes.
 * `AsyncWriteJournal`: A journal stores the sequence of messages sent to a persistent actor. An application can control which messages
are journaled and which are received by the persistent actor without being journaled. Journal maintains `highestSequenceNr` that is increased on each message.
The storage backend of a journal is pluggable. The persistence extension comes with a "leveldb" journal plugin, which writes to the local filesystem.
Replicated journals are available as [Community plugins](http://akka.io/community/).

>  * *Snapshot store*: A snapshot store persists snapshots of a persistent actor's state. Snapshots are

Reviewer 3: I might prefer to keep the 'internal' just to retain actor private / internal state vs being exposed by design. WDYT?

Author: My motivation for avoiding "internal" was that I initially thought "internal" meant the Akka framework's internal state of an actor or something, rather than "user-land" internal state.  Maybe it's just me?

Reviewer 4: I think it's better to remove internal for the reasons mentioned.
